### PR TITLE
Add advertise-addr support

### PR DESCRIPTION
**Pull request by mavenugo**
With this change, all the auto-detection of the addresses are removed
from libnetwork and the caller takes the responsibilty to have a proper
advertise-addr in various scenarios (including externally facing public
advertise-addr with an internal facing private listen-addr)

Signed-off-by: Madhu Venugopal <madhu@docker.com>

### CODE REVIEW

**Comment by mavenugo** [comment]
ping @aboch @aaronlehmann

**Comment by mavenugo** [comment]
This replaces the need for local-address as well for https://github.com/docker/docker/pull/24237.

**Comment by mavenugo** [comment]
Just realized that the tunnel encryption requires the localAddress.
changing this PR as WIP.

File: drivers/overlay/ov_serf.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ func (d *driver) serfInit() error {
 
 	config := serf.DefaultConfig()
 	config.Init()
-	config.MemberlistConfig.BindAddr = d.bindAddress
+	config.MemberlistConfig.BindAddr = d.advertiseAddress
```
Comment (aaronlehmann) [inline drivers/overlay/ov_serf.go]: Is it correct to set this bind address to the advertise address?

Comment (mavenugo) [inline drivers/overlay/ov_serf.go]: Yes. it is. Thats because ov_serf is not used in swarm-mode. It is used ONLY for the non-swarm-mode. In the non-swarm-mode, advertiseAddress is always the bindAddress and hence the above code is correct.

**Comment by aaronlehmann** [comment]
LGTM (not a maintainer)

**Comment by mavenugo** [comment]
ping @mrjana @aboch. Can you PTAL

File: drivers/overlay/overlay.go [hunk 2]
```diff
@@ -215,17 +216,18 @@ func validateSelf(node string) error {
 	return fmt.Errorf("Multi-Host overlay networking requires cluster-advertise(%s) to be configured with a local ip-address that is reachable within the cluster", advIP.String())
 }
 
-func (d *driver) nodeJoin(node string, self bool) {
+func (d *driver) nodeJoin(advertiseAddress, bindAddress string, self bool) {
 	if self && !d.isSerfAlive() {
-		if err := validateSelf(node); err != nil {
```
Comment (aboch) [inline drivers/overlay/overlay.go]: Why did you move down this validation check ?

**Comment by aboch** [comment]
I do not see any code which validates the bind and adv addresses.
At least before if a valid bind addr was  missing, libnetwork was trying to retrieve one via best effort.

Now, I think we should at least fail if we do not get a valid bind addr (and a valid adv addr if in agent mode)

File: drivers/overlay/overlay.go (hunk 2 above)
Comment (aboch) [inline drivers/overlay/overlay.go]: Never mind, this would fail in swarm mode when the adv address is not local to the machine.

File: drivers/overlay/overlay.go [hunk 3]
```diff
@@ -215,17 +216,18 @@ func validateSelf(node string) error {
...
-		}
 		d.Lock()
-		d.bindAddress = node
+		d.advertiseAddress = advertiseAddress
+		d.bindAddress = bindAddress
 		d.Unlock()
 
 		// If there is no cluster store there is no need to start serf.
 		if d.store != nil {
+			if err := validateSelf(advertiseAddress); err != nil {
```
Comment (aboch) [inline drivers/overlay/overlay.go]: To be correct, you should reset d.bindAddress and d.AdvertiseAddr in case of error, given now they are set unconditionally.

Comment (mavenugo) [inline drivers/overlay/overlay.go]: That is fine. We are not trying to change the existing functionality of external-store based overlay networking. am just trying to make sure it doesn't log an error for swarm-mode. It is just a warning to the user and we dont error out for this scenario even for the non-swarm mode.

Comment (mavenugo) [inline drivers/overlay/overlay.go]: Okay. resetting the values to "" even though it is not specific to this PR.

**Comment by mavenugo** [comment]
@aboch for your comment above https://github.com/docker/libnetwork/pull/1337#issuecomment-234135308, libnetwork is designed to handle the scenario of remote-addr, local-addr and advertise-addr to be valid across different calls to agentInit. Hence erroring out isnt appropriate and infact that is exactly designed not to error out on such scenarios.
But for ease of debugging, I added an Infof (which is useful even for the user to know which ip-addresses are chosen/configured for local-addr, advertise-addr and remote-addr.

**Comment by mavenugo** [comment]
@aboch addressed your comments. PTAL

File: agent.go [hunk 4]
```diff
@@ -236,25 +237,14 @@ func (c *controller) handleKeyChangeV1(keys []*types.EncryptionKey) error {
...
-		if err == nil {
-			bindHostPort := conn.LocalAddr().String()
-			bindAddr, _, _ = net.SplitHostPort(bindHostPort)
-			conn.Close()
-		}
-	}
-
-	if bindAddr != "" && c.agent == nil {
-		if err := c.agentInit(bindAddr); err != nil {
+	logrus.Infof("Initializing Libnetwork Agent Local-addr=%s Adv-addr=%s Remote-addr =%s", bindAddr, advAddr, remoteAddr)
```
Comment (aboch) [inline agent.go]: Thanks, this is good to have.

**Comment by aboch** [comment]
LGTM
